### DISCUSSION TITLE
Add Base URL to context

### DISCUSSION THREAD

**Discussion by chmcquee** [keycloakify/keycloakify#253] [Q&A]
Hi all,
I need the client baseURL set in keycloak from within the login screen.
Now all I get from keycloakify is the name , id and description of the client and not the baseURL.
Any ideas how to add this to the context?
Thanks :)

**Comment by chmcquee** [comment]
@garronej  I realised the login context doesnt have the baseUrl included. So I looked and found one in the info.ftl context.
But how do I get this value into my login context? :)

**Reply by garronej** [reply]
At runtime or only in types definition ?

**Reply by garronej** [reply]
It might be there even if not listed in types

**Comment by Fouad192** [comment]
Any solution for this issue?

**Reply by XINAD919** [reply]
This is how I created a go-back button that redirects to the project’s base URL:
```
const { auth, url, message, isAppInitiatedAction, client } = kcContext;

<a className="loginAccountBackClass" href={client.baseUrl}>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" fill="none">
         <path fill="#fff" fill-opacity="0.01" d="m12 19-7-7 7-7"></path>
         <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M19 12H5m0 0 7 7m-7-7 7-7"> 
         </path>
    </svg>
    <h4>{msg("doBack")}</h4>
</a>
```
To avoid type issues, I added baseUrl to the client object in the file ``` node_modules/keycloakify/login/KcContext/KcContext.d.ts```:
```
        client: {
            clientId: string;
            name?: string;
            baseUrl?:string;
            description?: string;
            attributes: Record<string, string>;
        };
```

**Reply by garronej** [reply]
Hello @XINAD919,  

Thank you for sharing your solution however you don't have to mess with the node modules to extend the type definition of the KcContext in your project: https://docs.keycloakify.dev/faq-and-help/some-values-you-need-are-missing-from-in-kccontext  

`src/login/KcContext.ts`
```ts
/* eslint-disable @typescript-eslint/ban-types */
import type { ExtendKcContext } from "keycloakify/login";
import type { KcEnvName, ThemeName } from "../kc.gen";

export type KcContextExtension = {
    themeName: ThemeName;
    properties: Record<KcEnvName, string> & {};
    client: {
          baseUrl?: string;
    }
};

export type KcContextExtensionPerPage = {
    // You can also define it at the page level: 
    /*
    "login.ftl": {
        client: {
            baseUrl?: string;
        }
    }
    */
};

export type KcContext = ExtendKcContext<KcContextExtension, KcContextExtensionPerPage>;


```

**Reply by Darfito** [reply]
hello I want to ask, I have used this method but I am confused why the url does not appear? do I have to fill the url also in package.json?

**Reply by garronej** [reply]
@Darfito The `kcContext.client.baseUrl` will only be defined if you have explicitly set it in your Keycloak client configuration. This corresponds to the "homeUrl" option:  

<img width="1134" alt="image" src="https://github.com/user-attachments/assets/d132b4dc-dd4d-418e-8806-5e116c07c774" />

<img width="866" alt="image" src="https://github.com/user-attachments/assets/5145669f-8f7d-4726-9ba0-f332c1095556" />

If you're looking for a reliable solution to implement a "home" or "back to app" button, you can first attempt to use `kcContext.client.baseUrl`. If it’s not defined, you can fall back to the origin of the `redirect_url` query parameter.  
The `redirect_url` indicates where the user will be redirected after a successful login, and its origin is a reasonable approximation for the homepage of the website.  

Here’s an example implementation:  

- [redirectUrlOrigin.ts](https://github.com/codegouvfr/keycloak-theme-dsfr/blob/main/src/login/shared/redirectUrlOrigin.ts)  
- [Template.tsx (line 80)](https://github.com/codegouvfr/keycloak-theme-dsfr/blob/bbe14c0e0c1a4b02e0bd75c3a57ab8e27e706a41/src/login/Template.tsx#L80)

**Reply by Darfito** [reply]
Well thank you! I really appreciate your answer, I will try it first and is it okay for me to ask further questions afterwards?

**Comment by garronej** [comment]
Base url is the url of your app? well, the subpath of the home?   
@Fouad192 tell me what you want to do I'll try and help you

**Comment by sreejith-kulamgarath** [comment]
This is how I did it when I wanted baseUrl

```
const {
        client,
    } = kcContext

const baseUrl = (client as any).baseUrl || ''

**Reply by garronej** [reply]
Hello @sreejith-kulamgarath,  

Yes, this is the quick and easy solution but there is a typesafe solution for this:  
https://github.com/keycloakify/keycloakify/discussions/253#discussioncomment-11055309
